bftengine/src/bftengine/PersistentStorageImp.cpp

```diff
@@ -0,0 +1,905 @@
+  uint32_t maxElementSize = DescriptorOfLastNewView::maxElementSize();
+  char *elementBuf = new char[maxElementSize];
+  for (uint32_t i = 0; i < numOfMessages; ++i) {
+    newDesc.serializeElement(i, elementBuf, maxElementSize, actualElementSize);
+    Assert(actualElementSize != 0);
+    metadataStorage_->writeInTransaction(LAST_NEW_VIEW_DESC + 1 + i, elementBuf, actualElementSize);
+  }
+  delete[] elementBuf;
+}
+
+void PersistentStorageImp::setDescriptorOfLastNewView(const DescriptorOfLastNewView &desc, bool init) {
+  if (!init) {
+    verifySetDescriptorOfLastNewView(desc);
+    verifyLastNewViewMsgs(desc);
+  }
+  saveDescriptorOfLastNewView(desc);
+}
+
+void PersistentStorageImp::setDescriptorOfLastNewView(const DescriptorOfLastNewView &desc) {
+  setDescriptorOfLastNewView(desc, false);
+  hasDescriptorOfLastNewView_ = true;
+}
+
+void PersistentStorageImp::initDescriptorOfLastNewView() {
+  DescriptorOfLastNewView desc;
+  setDescriptorOfLastNewView(desc, true);
+}
+
+void PersistentStorageImp::saveDescriptorOfLastExecution(const DescriptorOfLastExecution &newDesc) {
+  const size_t bufLen = DescriptorOfLastExecution::maxSize();
+  char *descBuf = new char[bufLen];
+  char *descBufPtr = descBuf;
+  size_t actualSize = 0;
+  newDesc.serialize(descBufPtr, bufLen, actualSize);
+  Assert(actualSize != 0);
+  metadataStorage_->writeInTransaction(LAST_EXEC_DESC, descBuf, actualSize);
+  delete[] descBuf;
+}
+
+void PersistentStorageImp::setDescriptorOfLastExecution(const DescriptorOfLastExecution &desc, bool init) {
+  if (!init)
+    verifyDescriptorOfLastExecution(desc);
+  saveDescriptorOfLastExecution(desc);
+}
+
+void PersistentStorageImp::initDescriptorOfLastExecution() {
+  DescriptorOfLastExecution desc;
+  setDescriptorOfLastExecution(desc, true);
+}
+
+void PersistentStorageImp::setDescriptorOfLastExecution(const DescriptorOfLastExecution &desc) {
+  setDescriptorOfLastExecution(desc, false);
+  hasDescriptorOfLastExecution_ = true;
+}
+
+/***** Windows handling *****/
+
+/***** Private functions *****/
+
+void PersistentStorageImp::serializeAndSaveSeqNumWindow(SharedPtrSeqNumWindow seqNumWindow) {
+  const size_t simpleParamsSize = sizeof(SeqNum);
+  char simpleParamsBuf[simpleParamsSize];
+  seqNumWindow.get()->serializeActiveWindowBeginning(simpleParamsBuf);
+  metadataStorage_->writeInTransaction(SEQ_NUM_WINDOW, simpleParamsBuf, simpleParamsSize);
+
+  char *winBuf = new char[SeqNumData::maxSize()];
+  for (uint32_t i = seqNumWindowFirst_; i < seqNumWindowLast_; ++i)
+    setSeqNumDataElement(i, winBuf, seqNumWindow);
+  delete[] winBuf;
+}
+
+void PersistentStorageImp::setSeqNumDataElement(const SeqNum &index, char *buf,
+                                                SharedPtrSeqNumWindow seqNumWindow) const {
+  SeqNumData &seqNumData = seqNumWindow.get()->getByRealIndex(index);
+  SeqNum shift = (index - 1) * numOfSeqNumWinParameters;
+  char *movablePtr = buf;
```

Comment:
How does movablePtr differ from buf?

Reply:
Allows delete of a right pointer in delete[].

Reply:
I'm not sure I really follow what you mean.

Reply:
When you serialize the data, a pointer is moved after addition of each serialized chunk to the buffer. But at the end, you should delete an original pointer and not one that points on the buffer end.

Reply:
The confusion is that the argument is not a pointer but a reference to a pointer which gets modified within the execution of this function: `SeqNumData::serializePrePrepareMsg(char *&buf)`. Hence, after we return from this function, `movabelPtr` has a different value than before.

Reply:
Why do you call it confusion? It is done intentionally. In case I need to serialize a whole element in the same buffer that's a way to do that.